common/rate-limiter/src/lib.rs

```diff
@@ -0,0 +1,318 @@
+            .interval(Duration::from_secs(3600));
+
+        let mut rate_limiter = builder.build();
+        assert_num_allowed(&mut rate_limiter, key_a, 2);
+
+        // Buckets can be refilled
+        rate_limiter.fill_throttles(1);
+        assert_num_allowed(&mut rate_limiter, key_a, 1);
+
+        // Buckets can be refilled more than one interval
+        rate_limiter.fill_throttles(2);
+        assert_num_allowed(&mut rate_limiter, key_a, 2);
+
+        // Requests larger than the number of tokens don't count, but are throttled
+        rate_limiter.fill_throttles(1);
+        assert!(rate_limiter.throttle(key_a, 2));
+        assert_num_allowed(&mut rate_limiter, key_a, 1);
+
+        // Buckets that can't be larger than the bucket size
+        rate_limiter.fill_throttles(5);
+        assert_num_allowed(&mut rate_limiter, key_a, 2);
+
+        // Other keys aren't affected by empty buckets
+        assert_num_allowed(&mut rate_limiter, key_b, 2);
+
+        // All buckets are filled at once
+        rate_limiter.fill_throttles(2);
+        assert_num_allowed(&mut rate_limiter, key_a, 2);
+        assert_num_allowed(&mut rate_limiter, key_b, 2);
+    }
+
+    #[test]
+    fn test_refilling_checks() {
+        let key_a = "Hello";
+        let default_bucket_size = 2;
+        let default_fill_rate = 1;
+        // Make interval long enough that it won't refill manually in this test
+        let builder = RateLimiterBuilder::new(default_bucket_size, default_fill_rate)
+            .interval(Duration::from_secs(3600));
+
+        let mut rate_limiter = builder.build();
+        assert_num_allowed(&mut rate_limiter, key_a, 2);
+
+        // We should not fill the bucket if the time hasn't passed
+        rate_limiter.check_refill(Duration::from_secs(0));
+        assert_num_allowed(&mut rate_limiter, key_a, 0);
+
+        // Filling the bucket should occur once if it's been the interval
+        rate_limiter.check_refill(Duration::from_secs(3600));
+        assert_num_allowed(&mut rate_limiter, key_a, 1);
+
+        // Filling the bucket should occur once if it's between 1-2 intervals
+        rate_limiter.check_refill(Duration::from_secs(3800));
+        assert_num_allowed(&mut rate_limiter, key_a, 1);
+
+        // Filling the bucket should occur once if it's been 2 intervals
+        rate_limiter.check_refill(Duration::from_secs(7200));
+        assert_num_allowed(&mut rate_limiter, key_a, 2);
+    }
+
+    #[test]
+    fn test_default_fill() {
+        let key_a = "Hello";
+        let default_bucket_size = 1;
+        let default_fill_rate = 1;
+        // Make interval long enough that it won't refill manually in this test
+        let builder = RateLimiterBuilder::new(default_bucket_size, default_fill_rate);
+
+        let mut rate_limiter = builder.build();
+        assert_num_allowed(&mut rate_limiter, key_a, 1);
+        sleep(Duration::from_secs(1));
```

Comment:
nit: one way to avoid having to sleep in tests is to pass in some sort of time service, and then in tests use a mock. If I remember correctly, we did this for the key manager: https://github.com/libra/libra/blob/1b32f757673bd16aff5cfd84901573b4eb1766f5/secure/key-manager/src/lib.rs#L105

Reply:
e.g., in the key manager tests we used, we applied a MockTimeService (https://github.com/libra/libra/blob/342116c954c6645fdaae192c4a448d7a46a08581/secure/key-manager/src/tests.rs#L363)

Reply:
I agree, should use a mock time service instead of actual time.  I'll fix it up so we I don't waste 1 second * the number of runs of developer time.